### PR TITLE
Implement `transfer_all` in Balances Pallet

### PR DESCRIPTION
**Pull request by shawntabrizi**
This extrinsic allows a user to send their full transferable balance, to a recipient account.

Can be helpful when trying to transfer a user's full balance, taking into account fees, and leaving no dust behind.

closes https://github.com/paritytech/substrate/issues/8696

polkadot companion: https://github.com/paritytech/polkadot/pull/3184

### CODE REVIEW

File: frame/balances/src/benchmarking.rs [hunk 1]
```diff
@@ -176,6 +176,22 @@ benchmarks_instance_pallet! {
...
+	// Benchmark `transfer_all` with the worst possible condition:
+	// * The recipient account is created
+	// * The sender is killed
+	transfer_all {
+		let caller = whitelisted_caller();
+		let recipient: T::AccountId = account("recipient", 0, SEED);
+		let recipient_lookup: <T::Lookup as StaticLookup>::Source = T::Lookup::unlookup(recipient.clone());
+
+		// Give the sender account max funds, thus a transfer will not kill account.
+		let _ = <Balances<T, I> as Currency<_>>::make_free_balance_be(&caller, T::Balance::max_value());
```
Comment (apopiak) [inline frame/balances/src/benchmarking.rs:189]: This is really confusing to me.
What do you mean with "thus a transfer will not kill account"? Isn't that the point of `transfer_all`, to kill the sender?
Do you mean to make sure that the recipient is not killed?

File: frame/balances/src/lib.rs [hunk 2]
```diff
@@ -369,6 +369,26 @@ pub mod pallet {
...
+		/// # <weight>
+		/// - O(1). Just like transfer, but reading the user's transferable balance first.
+		/// #</weight>
+		#[pallet::weight(T::WeightInfo::transfer_all())]
+		pub fn transfer_all(
+			origin: OriginFor<T>,
+			dest: <T::Lookup as StaticLookup>::Source,
+			keep_alive: bool,
+		) -> DispatchResultWithPostInfo {
+			use fungible::Inspect;
```
Comment (apopiak) [inline frame/balances/src/lib.rs:384]: Why is this needed?

File: frame/balances/src/lib.rs [hunk 3]
```diff
@@ -369,6 +369,26 @@ pub mod pallet {
 			<Self as Currency<_>>::transfer(&transactor, &dest, value, KeepAlive)?;
 			Ok(().into())
 		}
+
+		/// Transfer the entire transferable balance from the caller account.
+		///
```
Comment (apopiak) [inline frame/balances/src/lib.rs:374]: maybe add a note on what `keep_alive` does and how it leads to only a partial transfer?

**Review by apopiak — APPROVED** [review]
Could be merged as-is, I guess, but see comments above.

File: frame/balances/src/benchmarking.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/balances/src/benchmarking.rs:189]: yeah this is copy paste, sorry

File: frame/balances/src/lib.rs (hunk 2 above)
Comment (shawntabrizi) [reply frame/balances/src/lib.rs:384]: It was complaining I could not call `::reducible_balance()` without this import.

File: frame/balances/src/lib.rs [hunk 4]
```diff
@@ -369,6 +369,33 @@ pub mod pallet {
...
+		/// The dispatch origin of this call must be Signed.
```
Comment (emostov) [inline frame/balances/src/lib.rs:375]: Maybe not appropriate for here, but should there be some guidance to users of extrinsics like this on how provider/consumer ref counters may affect an attempt to reap an account?

I personally don't fully understand the use case of them.

In this case it looks like in `Balances::reducible_balance` we check
```
if frame_system::Pallet::<T>::can_dec_provider(who) && !keep_alive 
```
So as a user I would like to know in what situations  I _can't_ decrement the provider ref

File: frame/balances/src/tests.rs [hunk 5]
```diff
@@ -965,6 +965,46 @@ macro_rules! decl_tests {
...
+					assert_eq!(Balances::total_balance(&2), 200);
+
+					// setup
+					assert_ok!(Balances::set_balance(Origin::root(), 1, 200, 10));
+					assert_ok!(Balances::set_balance(Origin::root(), 2, 0, 0));
+					// transfer all and keep alive w/ reserved
+					assert_ok!(Balances::transfer_all(Some(1).into(), 2, true));
+					assert_eq!(Balances::total_balance(&1), 100);
+					assert_eq!(Balances::total_balance(&2), 110);
+				});
```
Comment (emostov) [inline frame/balances/src/tests.rs:1005]: Might be nice to add a test where we attempt a `transfer_all` with allow death, but the transactor has an active provider.

**Review by emostov — APPROVED** [review]
Overall looks good, just a couple comments

File: frame/balances/src/tests.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/balances/src/tests.rs:1005]: i think such a test should already exist with the `transfer` function.

if you see the implementation, all I am doing is calling the `transfer` api, but reading the "value" to send rather than have the user input it. So any behavior like this should be correctly handled by the API, or else we would have problems in every transfer function here in the balances pallet :)

File: frame/balances/src/lib.rs (hunk 4 above)
Comment (shawntabrizi) [reply frame/balances/src/lib.rs:375]: this function will not always reap an account. It will simply attempt to transfer as much transferable balance as possible.

Indeed other systems like democracy or staking (locks, reserves, ref counters) can affect a users access to transfer their full balance, but there is nothing that can be done about this at the pallet level.

Think about this tool as a perfect thing for exchanges to use when you send a balance to their "temp wallets".

I send some amount from A -> B, and want to send the full B -> C right after.

B will have nothing on the account except this balance which it wants to transfer, and this function is perfect for that.

Comment (emostov) [reply frame/balances/src/lib.rs:375]: My concern is around confusion for when a user tries to use this to reap an account, but they only send `free - ED` bc they have a `consumer > 0`. (I suspect having to deal with refs on an account is a new paradigm for the average user?).

So I guess my concrete suggestion here would be to elaborate on what qualifies as `the entire transferable balance`. i.e

```
// Transfer the entire transferable balance from the caller account. 
// NOTE: Valid transferrable balance does not include `AcountInfo.reserved` funds. Further, if the 
// caller has `AcountInfo.consumer > 1`, only transfer `AcountInfo.free - exstential_deposit` will be transferred.
```

File: frame/balances/src/tests.rs (hunk 5 above)
Comment (emostov) [reply frame/balances/src/tests.rs:1005]: Good to know - that def makes sense. 
One argument though for adding a test could just be to show edge cases of expected behavior to a user.

File: frame/balances/src/lib.rs (hunk 4 above)
Comment (shawntabrizi) [reply frame/balances/src/lib.rs:375]: I will add a bit more docs here, but by definition, "transferable balance" is defined as the balance which is allowed to be transferred.

This means no reserved balance, no locked balance, and no ED if outstanding references exist.

But yes, clarification of these points is worth having.

**Review by kianenigma — APPROVED** [review]

File: frame/balances/src/lib.rs [hunk 6]
```diff
@@ -372,9 +372,22 @@ pub mod pallet {
 
 		/// Transfer the entire transferable balance from the caller account.
 		///
-		/// # <weight>
+		/// NOTE: This function is only attempt to transfer _transferable_ balances. This means that
+		/// any locked, reserved, or existential deposits (when `keep_alive` is `true`), will not be
+		/// transferred by this function. To ensure that this function results in a killed account,
+		/// you might need to prepare the account by removing any reference counters, storage
+		/// deposits, etc...
```
Comment (apopiak) [inline frame/balances/src/lib.rs:379]:
```suggestion
		/// NOTE: This function only attempts to transfer _transferable_ balances. This means that
		/// any locked, reserved, or existential deposits (when `keep_alive` is `true`), will not be
		/// transferred by this function. To ensure that this function results in a killed account,
		/// you might need to prepare the account by removing any reference counters, storage
		/// deposits, etc...
```

**Comment by shawntabrizi** [comment]
bot merge
